Adding decision review

mdbenjam opened the pull request:
Added the decision review page at /dispatch/establish-claim/:id/review and moved the form page to /dispatch/establish-claim/:id/new.

**Test Plan**
- [x] Verify the new page looks as expected
<img width="1110" alt="screen shot 2016-11-27 at 12 05 24 pm" src="https://cloud.githubusercontent.com/assets/3885236/20649981/cdf73556-b499-11e6-8ae8-d4c2b9074320.png">

- [x] Add new unit and feature tests
- [x] Run test suite


**To Do**

- [ ] Ensure decision date is +/- 1 day of VACOLS decision date. -> This will be completed in a separate PR since it involves some large-ish backend changes.
- [x] Add Veteran name (First and Last) and Case name at the top of the page.

CC: @joofsh @sharonwarner

mdbenjam commented on app/models/document.rb:
```diff
@@ -31,7 +33,20 @@ def self.from_vbms_document(vbms_document)
+  end
+
+  def default_path
+    File.join(Rails.root, "tmp", "pdfs", "#{type.gsub(" ", "-").downcase}-#{document_id}.pdf")
```
Maybe this ending shouldn't be .pdf.

joofsh commented on app/models/appeal.rb:
```diff
@@ -29,7 +29,8 @@ def ssoc_dates
   attr_writer :documents
   def documents
-    @documents ||= []
+    fetch_documents!
```
Won't this call `fetch_documents` and make the API request every time? If we're caching them in an instance variable we should harness that.  I think something like `@documents || fetch_documents!` would work

mdbenjam replied:
fetch_documents! Makes sure that @documents is nil before loading it. So it should only call it once. We could move this logic, but I think the only advantage of that would be if we think we might call fetch_documents to reload the docs.

shanear replied:
I agree with you Mark. We should move the logic, its kind of confusing that the memoization takes place inside of `fetch_documents!`

joofsh commented on app/models/appeal.rb:
```diff
@@ -102,11 +103,21 @@ def missing_certification_data?
+    if (decisions.size > 1)
+      fail "Multiple decisions"
+    end
+    decisions[0]
```
What if there is no decision? This would return `nil`. Is that okay?

joofsh commented on app/models/appeal.rb:
```diff
@@ -102,11 +103,21 @@ def missing_certification_data?
   end
 
   def fetch_documents!
+    return unless @documents == nil
```
Oh I see now. So my problem with this approach is it's mixing responsibilities. `fetch_documents!` formerly was strictly responsible for making the API request and fetching the documents, and had no awareness of any sort of caching mechanism. We were able to "bust" the cache by calling `fetch_documents!` whenever we wanted. Now that functionality is removed if we check the cache here

joofsh commented on app/models/document.rb:
```diff
@@ -31,7 +33,20 @@ def self.from_vbms_document(vbms_document)
+  end
+
+  def save!
+    IO.binwrite(default_path, content)
```
This is pretty cool, and it looks like it'd work.  However I _believe_ it's still more idiomatic to use Ruby's `File` class (maybe check with Shane). It's a thin wrapper around IO specifically for manipulating Files. 
```
File.write('/path/to/file', 'Some glorious content')
```
Also just to check my own understanding, what is the `content` returned from `Appeal.repository.fetch_document_file(self)`? Is it the string contents of the document?

mdbenjam replied:
File inherits from IO, so it should have access to binwrite. I changed it to keep things consistent since I'm using File in other parts of the code. I think we need to use binwrite vs write, since the content we get is not a string, but just the binary format of the pdf. At least I'm expecting VBMS to return a binary version of the document, and that's what the test is doing.

mdbenjam commented on app/models/appeal.rb:
```diff
@@ -102,11 +103,19 @@ def missing_certification_data?
     [nod_date, soc_date, form9_date].any?(&:nil?)
   end
 
+  # TODO: (mdbenjam): Can there be multiple decisions?
```
Want to call this out. What should we do in the case of multiple decisions. Is that even possible?

shanear replied:
It is possible, because each eFolder can contain multiple appeals. I believe we need to match (exactly?) the decision by date. We need to ask @abbyraskin or @lakohl.

shanear commented on spec/rails_helper.rb:
```diff
@@ -10,6 +10,7 @@
 require "react_on_rails"
 require_relative "support/fake_pdf_service"
 require_relative "support/sauce_driver"
+require "react_on_rails"
```
double require lol?

shanear commented on app/models/appeal.rb:
```diff
@@ -106,11 +107,19 @@ def missing_certification_data?
+  def decision
+    decisions = documents_with_type("BVA Decision")
+    fail "Multiple decisions" if decisions.size > 1
+    decisions[0]
```
nitpick, but `decisions.first` is probably more idiomatic here

shanear commented on app/services/appeal_repository.rb:
```diff
@@ -197,6 +197,18 @@ def self.fetch_documents_for(appeal)
     appeal
   end
 
+  # TODO: (mdbenjam): This is copied from efolder, any way to unify them?
```
good call. can we make this a tech-improvement issue in caseflow instead of a TODO

mdbenjam replied:
https://github.com/department-of-veterans-affairs/caseflow/issues/504

shanear commented on app/views/tasks/worker_index.html.erb:
```diff
@@ -2,6 +2,7 @@
   <div class="usa-width-one-whole task-start-wrapper">
     <% if next_unassigned_task %>
       <div class="task-start-cta">
+        <!-- TODO: (mdbenjam) fix button to be grayed out when there is no next task -->
```
This seems like it should already work. You have `<%= loading_indicator %>` and the init code for it initializes in the layout here:
https://github.com/department-of-veterans-affairs/caseflow/blob/master/app/assets/javascripts/application.js#L25

mdbenjam replied:
It's not a matter of the loading indicator, since the tasks are gathered every night at midnight. So we can't load new ones on the spot. We just need to display a message if there are no more tasks.

shanear commented on spec/feature/establish_claim_spec.rb:
```diff
@@ -65,13 +65,14 @@
       click_on @task.start_text
-      expect(page).to have_current_path("/dispatch/establish-claim/#{@task.id}")
+      expect(page).to have_current_path("/dispatch/establish-claim/#{@task.id}/review")
+      expect(page).to have_content("Review Decision")
```
maybe add them clicking cancel here too?

mdbenjam replied:
Added.

shanear commented on app/models/document.rb:
```diff
@@ -31,7 +33,20 @@ def self.from_vbms_document(vbms_document)
+  end
+
+  def save!
+    File.binwrite(default_path, content)
```
We want to cache this doc in to S3. I believe that requires us to setup S3 with caseflow, so we may want to make a new PR for that.

shanear replied:
or make a tech-improvement story for it 😄

shanear commented on spec/models/document_spec.rb:
```diff
@@ -42,4 +42,46 @@
+    let(:document) { Document.new(type: "NOD") }
+
+    it "lazy loads document content" do
+      expect(Fakes::AppealRepository).to receive(:fetch_document_file)
```
These tests should test that `#content` actually returns what `Fakes::AppealRepository.fetch_document_file` returns. Not just that it was called.

I believe you can use:
`Fakes::AppealRepository).to receive(:fetch_document_file).and_return("some great content")`

shanear commented on app/controllers/tasks_controller.rb:
```diff
@@ -10,16 +10,25 @@ def index
+  def pdf
+    decision = task.appeal.decision
+    return redirect_to "/404" if decision.nil?
+    decision.save!
```
This seems like it's always fetching the doc from VBMS. It might be better to retrieve from the filesystem if it's already downloaded.

mdbenjam replied:
Good call. Side note, should we have a way to periodically clean out the tmp directory?

shanear replied:
Everytime the app is deployed the tmp directory is wiped. But we also have a job in efolder express that does that. Maybe we should copy it over

shanear commented on app/controllers/tasks_controller.rb:
```diff
@@ -10,16 +10,25 @@ def index
     # Future safeguard for when we give managers a show view
     # for a given task
     task.start! if current_user == task.user
+    render "review_decision"
```
why `"review_decision"` and not just `"review"`?

mdbenjam commented:
This got much larger in scope than I'd planned. I've addressed all the comments and want to get this merged. But there is still a list of things that need to be done.

- [ ] Two tech improvement tasks #503 and #504 
- [ ] Settle on the approach for cases we can't handle
- [ ] Think about moving this page to react
